api/lib/polkadot-utils.js

```diff
@@ -0,0 +1,95 @@
+function extractState(
+  stashId,
+  slashes,
+  nominees,
+  activeEra,
+  submittedIn,
+  exposures
+) {
+  // chilled
+  const nomsChilled = nominees.filter((_, index) => {
+    if (slashes[index].isNone) {
+      return false
+    }
+
+    const { lastNonzeroSlash } = slashes[index].unwrap()
+
+    return !lastNonzeroSlash.isZero() && lastNonzeroSlash.gte(submittedIn)
+  })
+
+  // first a blanket find of nominations not in the active set
+  let nomsInactive = exposures
+    .map((exposure, index) =>
+      exposure.others.some(({ who }) => who.eq(stashId))
+        ? null
+        : nominees[index]
+    )
+    .filter((inactiveId) => !!inactiveId)
+
+  // waiting if validator is inactive or we have not submitted long enough ago
+  const nomsWaiting = exposures
+    .map((exposure, index) =>
+      exposure.total.unwrap().isZero() ||
+      (nomsInactive.includes(nominees[index]) &&
+        activeEra.sub(submittedIn).lten(2))
+        ? nominees[index]
+        : null
+    )
+    .filter((waitingId) => !!waitingId)
+    .filter((nominee) => !nomsChilled.includes(nominee))
+
+  // filter based on all inactives
+  const nomsActive = nominees.filter(
+    (nominee) =>
+      !nomsInactive.includes(nominee) && !nomsChilled.includes(nominee)
+  )
+
+  // inactive also contains waiting, remove those
+  nomsInactive = nomsInactive.filter(
+    (nominee) =>
+      !nomsWaiting.includes(nominee) && !nomsChilled.includes(nominee)
+  )
+
+  return {
+    nomsActive,
+    nomsChilled,
+    nomsInactive,
+    nomsWaiting
+  }
+}
+
+async function useInactives(api, stashId, nominees) {
+  const indexes = await api.derive.session.indexes()
+  let nominationsOverview
+
+  if (nominees && nominees.length && indexes) {
+    await api.queryMulti(
```

Comment:
why not use a Promise.all? ^^

Reply:
I should have read this before deleting the file :sweat:

Reply:
ah, recovered i!

Well, there are two ways I tried and none of them worked. Maybe you can shed some light here:

1st way:

```
const [optNominators, exposuresAndSpans] = await Promise.all(
            api.query.staking.erasStakers
              ? nominees.map( async (id) => await api.query.staking.erasStakers(indexes.activeEra, id))
              : nominees.map( async (id) => await api.query.staking.stakers(id) ),
              nominees.map(async (id) => await api.query.staking.slashingSpans(id))
        )
``` 

returns: 

```
OPT NOMINATORS {total: 0, own: 0, others: []}
EXPOSURES AND SPANS {total: 24820263773466674, own: 2323039621094, others: [{who: 15rWBwqqQvk9AyZFEQwkRpFfHjRur6k649zrD1xMzTRLRxau, value: 14028847105668}, {who: 138vpvFZb14Si5xVvLYDyDKLkCcC85Pgae8Nru5ABiHxffMb, value: 61000000000000}, {who: 13b9oETrz8au7tUm3t2P4kwV52LNbKRtB6Mq2eyaDf3ceSCB, value: 64000000000000}, {who: 13WFBuQcYbbEsSDAbBgAkSAzsa8UQXCcKQuU4PWJLanyb82N, value: 6534441951820}, {who: 13zLCfnA8kqShQrojhDQDT8EeRai17vVPW4W8442KKQ4HcjB, value: 34142730017825}, {who: 13Bza2tDZsMbcCDZaQiykm5Z8jpAthWtK8PHuFAK3W18AXsY, value: 328576145232223}, {who: 16Q1ZNjWpJwpun2hW7VThcird8KuBgEmBB8jJ2gNGcC2zZzQ, value: 15086557134825}, {who: 12bjECpFsZonaK3MZ3G2QnhUfiWtwme4EeaaETHA4aM8j6aE, value: 8066377383917}, {who: 14BLAvk253QwofRZYo4PfHXuvXvgjydY6Nici1hKLn9p35B9, value: 3499538515245}, {who: 15MU4F6KMprfJH4LHGHFzNLBAB1K6yUVyyFay2ukAuFkmXjg, value: 41973089925451}, {who: 14DMx1DiyZ8gpWvLuz7A7PESYCaWVGY5Z24La8aMQXvFQkmk, value: 5420321439307157}, {who: 1jzXXnBhZ8JgdiPVkMBQpUQaAuWbe77MUW2hrCbM9rztS2b, value: 1767769849741093}, {who: 153Z4v5HTvmPQ6WtUf3g7HtWBGyAaaM567KPRLZaJ1szAsp5, value: 10069302421525}, {who: 15ddXK1J1Ctnt2gUWRmjrx438xSUwZQW7tRFvcS1ikUoNBWJ, value: 4911998099355}, {who: 14smFRvupyvBB1F8zn94rFbrh8PemJeY8cibmTHnn8qKy98, value: 137432000000000}, {who: 121qDKDEX1goZR7441wR1YSmFeDWwo238p6VNuAMsujKSz6D, value: 39946982613}, {who: 1397R5TJM2mfeLcueoWc8ZsZdPapDJp5naCRMK3kBZ3qZevc, value: 384097031091819}, {who: 1Ci2Vf2zo5jicv32AbEp49QSvikFEbHecjepsm13wWAbbLP, value: 25021254003552}, {who: 14dftZ3S9i8qzhS8fdxDAqSGHtcgnhmsQCZLYSUc1vQYNNHj, value: 1005887917450}, {who: 16hUyFhBoLAbZLpjVnGJCj63KSaUF23buYK9NEw8xzfQ4z3X, value: 657576226249186}, {who: 13Z8XjXNEutVqRrwzfcGhd6FvhyhDpvgfzirCSFbf8wYDpGc, value: 9374332582272}, {who: 1s7ZDWomtsA5yU3TC9UpnwdmHo1XANTa4ALpZNrs9Fh83yB, value: 152844296099651}, {who: 13EjL9Urz2DHhR3VT1PPwQype1mhVciBsKyFT8WCwcXd3oxY, value: 691568216044083}, {who: 1W6f6T5mxz2ATndjakXBRvDzX36B7eDurbHEcDKMhnRH2Zx, value: 10039989623865110}, {who: 14Jh6QHn9pnbkVLfTUTaj3xk7EN2McbkevDYcVWrAxS1KsFJ, value: 4939011602173740}]}

```

2nd Way:

```
        const [optNominators, ...exposuresAndSpans] = await Promise.all(
            nominees.map(async nominator => {
                (api.query.staking.erasStakers
                ? nominees.map( async (id) => await api.query.staking.erasStakers(indexes.activeEra, id))
                : nominees.map( async (id) => await api.query.staking.stakers(id) ))
                .concat(
                    nominees.map(async (id) => await api.query.staking.slashingSpans(id))
                )
            })
        )
```

returns two weird functions.

Should return

```
OPT NOMINATORS
EXPOSURES AND SPANS [{total: 0, own: 0, others: []}, {total: 24820263773466674, own: 2323039621094, others: [{who: 15rWBwqqQvk9AyZFEQwkRpFfHjRur6k649zrD1xMzTRLRxau, value: 14028847105668}, {who: 138vpvFZb14Si5xVvLYDyDKLkCcC85Pgae8Nru5ABiHxffMb, value: 61000000000000}, {who: 13b9oETrz8au7tUm3t2P4kwV52LNbKRtB6Mq2eyaDf3ceSCB, value: 64000000000000}, {who: 13WFBuQcYbbEsSDAbBgAkSAzsa8UQXCcKQuU4PWJLanyb82N, value: 6534441951820}, {who: 13zLCfnA8kqShQrojhDQDT8EeRai17vVPW4W8442KKQ4HcjB, value: 34142730017825}, {who: 13Bza2tDZsMbcCDZaQiykm5Z8jpAthWtK8PHuFAK3W18AXsY, value: 328576145232223}, {who: 16Q1ZNjWpJwpun2hW7VThcird8KuBgEmBB8jJ2gNGcC2zZzQ, value: 15086557134825}, {who: 12bjECpFsZonaK3MZ3G2QnhUfiWtwme4EeaaETHA4aM8j6aE, value: 8066377383917}, {who: 14BLAvk253QwofRZYo4PfHXuvXvgjydY6Nici1hKLn9p35B9, value: 3499538515245}, {who: 15MU4F6KMprfJH4LHGHFzNLBAB1K6yUVyyFay2ukAuFkmXjg, value: 41973089925451}, {who: 14DMx1DiyZ8gpWvLuz7A7PESYCaWVGY5Z24La8aMQXvFQkmk, value: 5420321439307157}, {who: 1jzXXnBhZ8JgdiPVkMBQpUQaAuWbe77MUW2hrCbM9rztS2b, value: 1767769849741093}, {who: 153Z4v5HTvmPQ6WtUf3g7HtWBGyAaaM567KPRLZaJ1szAsp5, value: 10069302421525}, {who: 15ddXK1J1Ctnt2gUWRmjrx438xSUwZQW7tRFvcS1ikUoNBWJ, value: 4911998099355}, {who: 14smFRvupyvBB1F8zn94rFbrh8PemJeY8cibmTHnn8qKy98, value: 137432000000000}, {who: 121qDKDEX1goZR7441wR1YSmFeDWwo238p6VNuAMsujKSz6D, value: 39946982613}, {who: 1397R5TJM2mfeLcueoWc8ZsZdPapDJp5naCRMK3kBZ3qZevc, value: 384097031091819}, {who: 1Ci2Vf2zo5jicv32AbEp49QSvikFEbHecjepsm13wWAbbLP, value: 25021254003552}, {who: 14dftZ3S9i8qzhS8fdxDAqSGHtcgnhmsQCZLYSUc1vQYNNHj, value: 1005887917450}, {who: 16hUyFhBoLAbZLpjVnGJCj63KSaUF23buYK9NEw8xzfQ4z3X, value: 657576226249186}, {who: 13Z8XjXNEutVqRrwzfcGhd6FvhyhDpvgfzirCSFbf8wYDpGc, value: 9374332582272}, {who: 1s7ZDWomtsA5yU3TC9UpnwdmHo1XANTa4ALpZNrs9Fh83yB, value: 152844296099651}, {who: 13EjL9Urz2DHhR3VT1PPwQype1mhVciBsKyFT8WCwcXd3oxY, value: 691568216044083}, {who: 1W6f6T5mxz2ATndjakXBRvDzX36B7eDurbHEcDKMhnRH2Zx, value: 10039989623865110}, {who: 14Jh6QHn9pnbkVLfTUTaj3xk7EN2McbkevDYcVWrAxS1KsFJ, value: 4939011602173740}]}, {total: 24816735042341005, own: 10000000000, others: [{who: 16JSNGrqQKyx2pQur7TvqipVgPGRVBCRKUCCsJCxNcrunuN8, value: 143248241612162}, {who: 15YRqDv5mYuVoeFcW7Y2NWgB5Ut6bc9C8y35yAC7aT6P3YzD, value: 9662238340877071}, {who: 158NdmXq6vjedNKHqbxfCeh2GbkZPC96cCAGmYkRGwxHpjAB, value: 38467153501430}, {who: 12in7txJQBFYWuJCcuYsTFyxvavBsQPx2Z17VrfFSJYfpzPy, value: 58002329518496}, {who: 12UDW1A4NB2rmeCFjNqzR8oogKE4tWTgGszLD9dPrRB2N9ah, value: 3013554416669}, {who: 12VojFkh4XDhu1dPKWhkbaTT2QJUthTPSSSaGvVXnN51Wb2m, value: 111465392338621}, {who: 13DpWq1ECsJmoAN6y4j6S8Hmd2S4tmeD5vidZ7Pbre2CcKtG, value: 768256354526247}, {who: 12W3ncyAfWR2sdrbV4ud2asYxPjUnwUwL2xhKttNAdBjqJa3, value: 46086764790363}, {who: 14Fst5uCLEP9ZBkaXuaf5j6DToLFqzsU2XqLzYLxwLwpuEf3, value: 1196386066002}, {who: 16JCybAA88yQ9t8Cus4YhB5mT5DjyBxBLEgYPCpH8HjnePTq, value: 1050000000000000}, {who: 16UM3kaTMMfXfYJ2RxBcFJ5DreM7wV1cEVUwi6YE85SG8h1j, value: 19508166866743}, {who: 13QBnezwes4mDDdeftTzkpGN9iADkfCn3SvbD5LT9hu2yukJ, value: 183905045122512}, {who: 15qbbXYvdxq1NLyRdh4kkcLNsbeYbkzDXk4dJZtvJVkCTRL3, value: 13719224186811}, {who: 14rRSBG25Mtpd7YBpA2UGL97Ly3AFUkLMiKajUAztgAa5HLp, value: 100768564681822}, {who: 13uvqQ1zjWaJtbXgC9ssNX1jA7D27RUUoBMzUn33jypocE8Z, value: 7878580910963607}, {who: 14H1z5iMGrELPM31o2nbobZ4CVUuurtBbnzcEsbFQUJLM1Yk, value: 4738268612872449}]}, {total: 24813557432759510, own: 4073859188552, others: [{who: 1NgthSfo8mRdbxH77DRVcBpMjCABhZFThgeaoxgbymkSTuK, value: 2267398294025}, {who: 16G473WLPNYNBucgzQphdB7gEvS7RHFKRY51J8gf9kaJGNKc, value: 106391756991031}, {who: 133yK7vU4p24bGSnqyMTinY4Kfo8uHfgubhPFqEB82XxgC8N, value: 95512978955258}, {who: 162ed9noHDZtq9UESp82b9rjbmHZj6n5Fs2sK5Cv777hXgAs, value: 61000000000000}, {who: 16cgpcfX5uX3FjQX6SXWRw3NRdYk8nVBBbNXRRJbVkUD81sF, value: 300000000000000}, {who: 13zWtMxyDQ3FwT35jMnDymqRZ5xbQUWsJ3bSFX4rf7ChoJwd, value: 7685426698996}, {who: 1556LC1pFMiUWpxAN1FFqCBhHrHuGrkLCEctSnyqP74AFrtE, value: 17519413568170}, {who: 13pnKLerWjzYiQ41FimgFSrXU82ksXLdDAiQPW3JsXYpTbQL, value: 422597403880138}, {who: 12rwQRSYSByUfJg25RBzELAsFer7tExZv5MNbiBaPpJxJWgM, value: 1473804497302}, {who: 12QmRJJ1qkL9bm9LsUqBkwhYEUKqWLWB47nqyS3bphJ9N1CD, value: 152847417998932}, {who: 15E6nqy4aSHvPYgDB8wSyE3d34yYdqpAg4ikQGPHr7g4NG17, value: 3019206306033}, {who: 12BH9FxgtjJG8zk9boXsnGoBkYKuM1c6MPsTXAQJ5Fb1YZsK, value: 931181435748538}, {who: 1nRsCk2UiNcEQjoQDr2kSyEE9YXaxaBW7As4fapdxqE2zDh, value: 18106014764241}, {who: 14BBTBJjAvEASe8cgfTpr4m3a7HBngWLFSWPRWEoMoUoiFoR, value: 46194181741}, {who: 13Ynf6wVVpaZegURhv2MsTQMw5wPZ3i52Uuw5rtUZyEdRAvn, value: 2112381428514132}, {who: 16SZwouo3TjEZrfLQyzsC5SN5o55F1DnpEgTJPxLZR9inrDh, value: 4123019942039}, {who: 1HGxwRgJFfma2QmfkHC474v3Gw6LHZT9HfxRXxq5fuonStw, value: 160896816860943}, {who: 13zUaQSZdDzivuHAR4C3qUt7VisbEFjuZWP6HdaWVB7LsZnv, value: 606494975154968}, {who: 15dmcrSxy1Hi44A2iJQ6gqtVo3ZyFXqGM4PapaJEAxqcX6iZ, value: 98500649196703}, {who: 15dgRFFbH5RostnyCs6FmKir1LaKcWwjjgRJjUdMH4ymA7gL, value: 2252793032995}, {who: 16JyV9y1phkkHBFs5PXLQK4atWNDkreoabMbXN4aHpeZzw3S, value: 5028065687863}, {who: 16dr2uMBBaiFKYWQxEdoUAiEc1jSkVzfqmLx6QWrd5HxNZD, value: 22925785245344}, {who: 15fjimmGCjAAqVLo9o5DpAQMgrWYWRFXQixRqCXd8Xu1dxEe, value: 233668981419308}, {who: 1WbK3qvsZLKshdXZP4bhXUf7JTaFDmVXx1nmLtkUU62XtBf, value: 39579920123869}, {who: 1YyRQWnHf9rmeP7bymihmQgw1BXzSKsfsoVVWGKtiBjU1LN, value: 2270751661911361}, {who: 1y1GG1WtSiygitoFYuxnrtt3jo3XTVQxhcG2u9vvu2uR3ND, value: 17133231024597028}]}, {total: 0, own: 0, others: []}, {total: 33110673279258265, own: 5163975736397007, others: [{who: 12CFeee5mx2Chz77YWGZeEQKbCBsp1ayxfXAeJa77Aamg7hC, value: 967040153568740}, {who: 14T6JxuEQ7uUnvU6FyDi45buutzpHJScedCgf4X6xEgmjdCv, value: 2646636210722032}, {who: 164zbtB2XLoFtJWhuoX4PQjJvtJjEHgxQ8aRcfFpHtE8tq3e, value: 10124414897}, {who: 1sNDeNNeHTTmg76mNGAJN9hmew9KCS1JruadQr5A1pxWMbu, value: 77437341773107}, {who: 1Qk4qQ56Crm2ozBnBiL9rKinKADHKQk9xgKsmmrc6WK1DTM, value: 373535157361755}, {who: 1H4VuA6T9fb3xt6VM4mYaVquEyZ6Am7RzuYhyoFzi4hvjKN, value: 18313495192112}, {who: 14kpvbmYhtfFZ7BmXU6ejH754nSPXkcD35iWwY9HB6aSrC7e, value: 116262942683006}, {who: 152zrxtaNuvT96HWQN5Cw7Tbn8yNAE6qUNmoM47A476Fu4aG, value: 10124414897}, {who: 13C1z8Hh69qq9qsTaLWwVzeny258Z6UzeJsSAn4Zfy1xQucW, value: 1526722720127023}, {who: 16CfBoW4ZYPnwPG4EFjvzQaMnaBmNGP2rMB2D8ByBPEDLYHW, value: 437941820439637}, {who: 14Xs22PogFVE4nfPmsRFhmnqX3RqdrUANZRaVJU7Hik8DArR, value: 21782787452164052}]}, {total: 31234348621908362, own: 33094301488697, others: [{who: 15ziwhAnJJSHy4228q47k2Uo8wnyhMsztxKkYqC72Ccb5DZo, value: 81453230428603}, {who: 16iZfdma99cC9D5CAW8jN1Y9AQdZ65pGp6Tj2QaHNr8BmPZH, value: 813814474840932}, {who: 13Vka4qGSStrNoFZap9qryQCbubfjDVyeradJwU2BG7TxZir, value: 79411135116225}, {who: 14gV6s4whdfezXgnT6tJ79nCKLy2GutDpnEVSMBR9TmDpw5m, value: 2856855115587091}, {who: 12WnWGFpWNWUyn8uuahG2M8JzG82DEN8hoPi5QPptncEdWve, value: 29950108413447}, {who: 1tZzPmcq8Auisttygmg9g6tPMtrh9i3b22D3tKXvde7ibRB, value: 19020884017662966}, {who: 14AoK8VSrHFxZijXhZGSUipHzZbUgo2AkmEaviD9yxTb1ScX, value: 8318886238370401}]}, {total: 0, own: 0, others: []}, {total: 0, own: 0, others: []}, {total: 24833326340488364, own: 5000000000000000, others: [{who: 16QMadKUDxcfYJt2KEjg34jccFGXJFVpo6PKpRmJwUgJvgc4, value: 50000000000000}, {who: 1C3oJMnmAnu5eKyF2ZeJLwWjgf6UxauZb2GL5ZDGqc63QKu, value: 92620504671172}, {who: 15aGvWjHZeiAV2y6cam2FtPAnqFcq8n3Wtbpa6RZD36NawQJ, value: 12000000000000000}, {who: 13gVfEJg5sFGAvWSFA9G4N96yYdATyrTupxHUmTHaunrMZjz, value: 5117422153327626}, {who: 14aVp8yhWK2emdxbMP9U5rW2CZj3AuxhoAYQSkTNwXNCsQ1N, value: 4300970581882}, {who: 15f99C4YP46EDUBDhA2CkhxrKq7GatqgamkjtgykpiX2hJLt, value: 422544968668633}, {who: 15oEZrn5kLSkCUb4sHjPx5RHEBKJkKQuLMhEZajZ5cuALvnC, value: 102163979609641}, {who: 13uADiR1mkCiRr7RKs2oXUEEm6pa4n6WhxfaVn5u5dykXkvg, value: 102176315225138}, {who: 163HXb6fNGtbStL4t5dMinFMYSWjXiotw7yfkdgFXMd7gb1M, value: 1942097448404272}]}, {total: 30661070639891259, own: 4423101721494, others: [{who: 121irgcHTeP2JQPeKXEjsDE9Yd8Vrregr6gWqWdG9sFdkX6s, value: 5349833549155}, {who: 11gh7WpYe5CKN82opEkrZUJsHeaFZa3pdEkEZFSMkV3k1Xj, value: 202081156805186}, {who: 15FXTbSvd5wHm6NY95HpbC2rH1sjBGZuyGNY5gKjRLcEa8hx, value: 38099978899739}, {who: 15Akq5qx9keSR4N8QxgPe8Z2tF722KhQj6tt1dQkkj31dhZT, value: 26488206384525}, {who: 16K2MCLZzRVToYjzDxuqzq9KXbK8SSJieCGgMNcMkHwXigRH, value: 4015304522464}, {who: 1h8oqBDPjWWnHLHuoS3Q8SYHcx2BNYexzwJPMRJttGGyXFf, value: 235830635001}, {who: 13VMVEkQ2LNFF8XTQivn4v4nBfb3Ns4xPX1LFGP7DfzkrsRs, value: 4076692267923}, {who: 16CfBoW4ZYPnwPG4EFjvzQaMnaBmNGP2rMB2D8ByBPEDLYHW, value: 9761246950886457}, {who: 14UpRGUeAfsSZHFN63t4ojJrLNupPApUiLgovXtm7ZiAHUDA, value: 20615053584219315}]}, {total: 0, own: 0, others: []}, {total: 30553909893264178, own: 46614928852990, others: [{who: 15awRyNhY6fZS6Lg2wp4NLVeFH36vciU8gts2gshz4ierXhC, value: 1829597196998146}, {who: 14GD7cQdgZCoNYu6tkyMjj3ZzqHmLEBm8r3suadASeGFXG9R, value: 250367274903510}, {who: 16a88qvRh5y5uwB1NyuKrSDpFrMMhPGf97BwuVtezXWhcMPP, value: 2147753158484593}, {who: 16D5Wq792vQS8xwqycMTBKh3sXPrz8K8dj4CBqvWFybiqwJn, value: 55998393436198}, {who: 1NmeQXi2xPHkpR6kicgzQ625ik2LwfvG5wSHv2zufYGgRrh, value: 80434056175832}, {who: 14gV6s4whdfezXgnT6tJ79nCKLy2GutDpnEVSMBR9TmDpw5m, value: 26143144884412909}]}, {total: 0, own: 0, others: []}, {total: 0, own: 0, others: []}, {total: 0, own: 0, others: []}, , , , , , , , , , , , , , , , ]

```

Reply:
Yes, it is complex